Add Form Patterns Documentation

jorytindall opened the pull request:
### :pushpin: Summary

This PR adds the first round of documentation, guidelines, and recommendations around form patterns. This is a fairly exhaustive bit of content, but there are still several areas that I've captured to address in the future in the [epic](https://hashicorp.atlassian.net/browse/HDS-1407). This is the first initial round with many of the high-level topics covered.

didoo commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+
+### Text
+
+Sections consisting of text should use a 8px vertical gap between text elements.
```
I am not sure this is correct: if I look at the illustration above, I don't expect to see text-only sections. Maybe what we should say here is something like _"multiple* blocks of text within a section should use a 8px vertical gap..."_

(*) I think there is a better word, but I can't remember it, something like "subsequent" or something like that

jorytindall replied:
Yep, that's what I'm trying to say here, I'll update

didoo commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+
+#### Button grouping
+
+Buttons should be grouped using the [ButtonSet](/components/button-set) guidelines, e.g., using a 16px horizontal gap between each button.
```
```suggestion
Buttons should be grouped using the [ButtonSet](/components/button-set) guidelines, e.g. using a 16px horizontal gap between each button.
```

didoo replied:
There are also other similar occurrences in the text, so is clearly intentional. Maybe double check with @KristinLBradley (I am not sure I've ever seen it used this way)

jorytindall replied:
I believe this was how @heatherlarsen was using "e.g." when part of a sentence. Is this right? ([For reference](https://www.grammarly.com/blog/know-your-latin-i-e-vs-e-g/])

heatherlarsen replied:
I'm no grammar expert, but this is what Grammarly corrects it to, so that's what I've been using across the docs. The link @jorytindall shared has more details.

KristinLBradley replied:
I would go by Grammarly and the linked reference on this one

MelSumner replied:
Yes, the comma should stay there. 👍

didoo replied:
OK, it's an American English vs British English thing. All good then :)

<img width="677" alt="screenshot_2442" src="https://user-images.githubusercontent.com/686239/223101580-4f9482b7-5950-4219-a993-c4ada13e8fd4.png">

didoo commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+
+#### Technical and application needs
+
+Sometimes the values or options for a given field are dependent on a prior field or piece of information provided to the application. These technical contraints within the application are often expressed three conditional logic within the form; determining what is displayed to the user depending on what has already been selected or input in previous fields.
```
"three" ??

didoo replied:
also "depending" → "depends" ?

didoo commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,229 @@
+
+Only use dividers between sections of a form, not between fields or other smaller elements within a section. A divider should be used to break up different types of content and categories within a form.
+
+When using a divider, increase the spacing between sections to 48px from 32px (24px on the top and bottom of the divider). This helps to separate the divider from surrounding form elements, otherwise it can appear "attached" to a specific section or field.
```
Instead of saying "When using a divider, increase the spacing between sections" (which can be confusing) we can just say that the divider should have `24px` of margin above/below (it's the same thing, but it's simpler as a concept to understand)

didoo reviewed (commented):
Did a first initial pass (I imagine there will be more, right?)

Some comments ([already mentioned](https://github.com/hashicorp/design-system/pull/1224#pullrequestreview-1321796954) in the other PR about patterns for buttons):
- images should be optimized
- tabs organization can be different, if needed
- there's no information on the "code" side: as a developer, are there already components that I can use to achieve what we're suggesting to do in the docs? if not, what am I expected to do? if components will be created, will be made avaible soon?

jorytindall commented on website/docs/patterns/form-patterns/partials/specifications/examples.md:
```diff
@@ -0,0 +1 @@
+## Examples
```
Good catch, I meant to remove that

jorytindall commented on website/docs/patterns/form-patterns/index.md:
```diff
@@ -0,0 +1,17 @@
+---
```
Added an image, I think it makes sense to use the preview cards for pattern pages, but we'll probably need to rework more of the content on the patterns page/template.

jorytindall commented:
> (I imagine there will be more, right?)

Yes, I have more content to add and topics to include, but this is what is making it into the first round. I'll set up some time for us to discuss how this might be translated into code for developers (whether we want to build components or not)

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+#### Alignment
+
+Fields organized in a horizontal group should be aligned to the baseline of each element to account for fields which have helper text.
```
```suggestion
Fields organized in a horizontal group should be aligned to the baseline of each element to account for fields with helper text.
```

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+![Baseline alignment within a group](/assets/patterns/form-patterns/baseline-alignment.png =500x*)
+
+#### Multi-column layout
```
I see an opportunity to display visual do/don't blocks in this section to help consumers better comprehend this content. It can be a follow-up task, but I wanted to note it anyway because I think it would be a beneficial addition.

jorytindall replied:
Yep, that's a good idea. I added visuals and do/dont blocks here.

heatherlarsen reviewed (commented):
Lots of great content here! This is going to be so helpful for consumers ✨ 

I'm about halfway through reviewing this and will wrap up tomorrow, but I wanted to get some feedback over to you in the meantime. I left individual comments throughout, mostly grammar/brevity stuff, but I have two bigger pieces of general feedback:

- I find the different colored borders in the visuals hard to understand. There's no key or indication of what each border means, and I wonder if, for each image, you outline the specific piece you're trying to showcase with a consistent color (information-100 or success-200 maybe) and the rest of the pieces with a neutral color?
- the "Groups" section confuses me a bit because, in one example, you have a "group" of fields with one label (_legend_), but in the rest of the examples, the "groups" each have their own label. Is there a way to more clearly define when a group of fields should use a legend vs. using their own labels?

jorytindall commented:
> I find the different colored borders in the visuals hard to understand. There's no key or indication of what each border means, and I wonder if, for each image, you outline the specific piece you're trying to showcase with a consistent color (information-100 or success-200 maybe) and the rest of the pieces with a neutral color?

I was trying to differentiate a bit between the different semantic elements I outlined (form, section, group, etc). I'll take a pass at establishing more consistency.

> the "Groups" section confuses me a bit because, in one example, you have a "group" of fields with one label (legend), but in the rest of the examples, the "groups" each have their own label. Is there a way to more clearly define when a group of fields should use a legend vs. using their own labels?

This was a point raised by @didoo and @alex-ju as well, I have a work item [here](https://hashicorp.atlassian.net/browse/HDS-1605?atlOrigin=eyJpIjoiZWZlNTQ2YTMxMTlkNGUyN2FjOGM1ZTllYTA0N2UyOTUiLCJwIjoiaiJ9) for it that I plan to cover in our debrief next week

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+A form should hug the content and the fields it contains.
+
+![Form width](/assets/patterns/form-patterns/form-width-example.png)
```
Can we provide more realistic examples? Plopping a radio card example within a credit card form feels a little weird.

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+#### Field width
+
+The width of a field within a form should be wide enough to account for the estimated width of the content it accepts. This gives the user an accurate sense of the length and type of the content the field accepts and is important in setting user expectations.
```
```suggestion
Generally, the width of a field should be wide enough to account for the estimated width of the content it accepts. This gives the user an accurate sense of the character length and type of content the field accepts and is important in setting user expectations.
```

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+#### Single column layout
+
+If a form uses a single-column layout, consider setting a maximum width on the form; this can generally be best achieved by using a size or unit that is **relative** the page, viewport, or container size. This can commonly be achieved by using a viewport width unit (`vw`), percentage width (e.g., `50%`), or a character unit (`ch`).
```
Is there a reason we aren't recommending single-column layouts? It feels like this is an easy opportunity for us to gain alignment across products.

jorytindall replied:
It might make sense to recommend this as the default starting point and only to diverge if absolutely necessary. I'll reword some of the content here and reorganize it

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+!!! Info
+
+The web is a fluid medium; a relative width will adjust based on the viewport and container width. Designing and building a form with a recommended maximum and minimum width will help to conceptualize the graceful expansion and contraction of the viewport
```
```suggestion
The web is a fluid medium. A relative width will adjust based on the viewport and container width. Designing and building a form with a maximum and minimum width will ensure a graceful expansion and contraction within the viewport.
```

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+### Responsive properties
+
+As the viewport shrinks, the form width should expand relative to the viewport width, eventually occupying the entire width of the viewport or page. Horizontally grouped fields that result in a multi-column layout should stack vertically as the viewport shrinks.
```
I understand what you mean, but it took me a minute to wrap my head around it because it sounds backward. Wondering if a visual example could reduce confusion around this topic.

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+## Progressive disclosure
+
+Users want power, but also want simplicity. Progressive disclosure refers only displaying the most important information (or elements of a form) and conditionally showing additional fields depending on:
```
> Users want power

This feels like an odd statement to me. What are you really trying to convey here?

heatherlarsen replied:
```suggestion
Users want power but also simplicity. Progressive disclosure refers to only displaying the most important information (or elements of a form) and conditionally showing additional fields depending on:
```

jorytindall replied:
I have some beefing up to do on this content captured in [this ticket](https://hashicorp.atlassian.net/browse/HDS-1610), but essentially what I'm trying to say is progressive disclosure is a method that can be used to different levels of complexity within an experience depending on a number of different factors (settings the user has enabled, past behavior, customizable settings, etc).

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+1. the input state of other fields
+2. if more complex settings are requested by the user.
+
+## Required vs. optional fields
```
This is the heading used in the form component pages. 

```suggestion
## Required and optional
```

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+![Comparison of forms using optional vs required](/assets/patterns/form-patterns/required-optional-comparison.png)
+
+<!-- Validation -->
```
This content is already written on the [Form Primitives](https://helios.hashicorp.design/components/form/primitives#error-validation) page. While I'm sure the content that's there could be expanded on _(later)_, it makes sense to move it here so it's with the rest of the pattern documentation.

jorytindall replied:
Killed it _for now_

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+When determining whether to mark fields as optional or required, consider which method occurs the least. This can help reduce visual weight since unmarked fields are implied as the inverse of marked fields.
+
+![Comparison of forms using optional vs required](/assets/patterns/form-patterns/required-optional-comparison.png)
```
It feels like this visual would benefit from a do/don't distinction.

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+## Required vs. optional fields
+
+Displaying to the user what fields are required or optional can help users submit a form free of errors, but can also add visual weight the form.
+
+- For shorter forms fields should be interpreted as required by default. Fields that are not required should be marked as optional.
+- Don't mix required and optional labels, stick to one or the other. Using one method will imply that fields without a label are the inverse of whatever method you choose.
+
+When determining whether to mark fields as optional or required, consider which method occurs the least. This can help reduce visual weight since unmarked fields are implied as the inverse of marked fields.
```
This content is already available within the form component documentation (ref: [text input](https://helios.hashicorp.design/components/form/text-input#required-and-optional)) and can be used here directly. Let's ensure we're reusing existing content when it's available. 

This new guidance seems to differ slightly from the original guidance, so let's discuss it tomorrow.

heatherlarsen reviewed (commented):
I'm sorry for the insane number of comments! This felt easier than me going back to review and edit later. Again mostly small grammar/brevity things and also a few re-organization suggestions. 

Plus one new general observation:
- The hierarchy for "Order and organization" feels a bit off. 
  - I think it's accidentally nested a level too deep, so I provided suggestions to fix that
  - I noticed there are 5 headings under "Order and organization" but in the overview paragraph, you only mention forms falling into three categories. Wondering if adding a heading above the 3 types could help...something like this:
<img width="271" alt="Screenshot 2023-03-07 at 2 08 29 PM" src="https://user-images.githubusercontent.com/8553306/223564878-536329ea-a4cb-4b7c-bf85-66244685105b.png">

jorytindall commented on website/docs/patterns/form-patterns/partials/specifications/anatomy.md:
```diff
@@ -0,0 +1,19 @@
+| ------- | ----------------------------------------------------------------------- |
+| Form    | Consists of one or more sections                                        |
+| Section | Consists of multiple content types (text, fields) and groups of content |
+| Group   | Layout mechanism to group like elements together                        |
```
I think this is going to be a common question, so I'll try to add informational content wherever this seems relevant to note that we don't publish any form helper components related to layout, etc.

jorytindall commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+### Form
+
+A form wraps the content, fields, and actions while simultaneously handling the logic for submitting and action actions performed by the form. At a fundamental level, a form also acts as a layout mechanism, adding consistent spacing between sections.
```
I took another stab at the content here to clean it up and add some context. I think it's important to keep this in as it validates the form as both something functional (handling the logic) as well as handling layout and spacing.

jorytindall commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+A **field** is a broad term used to describe a form control and label pairing. When combined within a section there should be a 24px gap between each field.
+
+![Example of multiple fields in a section](/assets/patterns/form-patterns/section-fields.png =450x*)
```
I'll include this update in cleaning up the rest of the images to use a more consistent color palette

jorytindall commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+#### Text fields
+
+<!-- TODO -->
```
I'm going to remove this for now as it's still WIP and captured for the future via [this ticket](https://hashicorp.atlassian.net/browse/HDS-1606?atlOrigin=eyJpIjoiN2EwNjYwMjU2ZTQ5NDY0MDgxNDg3ZTM1M2QwZjAzODIiLCJwIjoiaiJ9)

jorytindall commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+![Example of a text section](/assets/patterns/form-patterns/text-section.png =600x*)
+
+Text elements within a form should use logical, step-based sizing to reinforce hierarchy within sections, as well as within the form itself. While specifics around type hierarchy should be determined at the application level, adhering to these guidelines will help establish consistency both at the page level, and when constrained within another element or component.
```
This is the challenge that I've had with the content, it does mention spacing as it relates to text elements, but also mentions other elements when using text within a form. Maybe it makes sense to bubble it up higher to an `h2`

jorytindall commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+#### Business needs
+
+<!-- TODO -->
```
I've removed them for now since the follow-ups are captured in a ticket

jorytindall commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+### Order and organization
+
+Fields within a form generally fall into three categories that can help determine order and organization within a form. These categories don't exist in isolation but rather have a cumulative effect on how a user perceives a form and how likely they are to complete it.
```
Yep, makes sense!

jorytindall commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,236 @@
+
+#### Form width
+
+A form should hug the content and the fields it contains.
```
Killed it

jorytindall commented:
@heatherlarsen several updates captured here:

- Changed "group" terminology to "set" and cleaned up the language where relevant
- Fixed a couple of typos
- Added cross-linking in the related section of form components to pattern docs

Validation:

- Moved this content from primitives to form patterns and relinked references in the various form component pages.
- Cleaned up the content section in the `RadioCard` page to point to the primitives section on content which is what all other form components were doing (the content was duplicated on the `RadioCard` page except for a banner specific to the component)

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,269 @@
+
+Text elements within a form should use logical, step-based sizing to reinforce hierarchy within sections and the form itself. While specifics around type hierarchy should be determined at the application level, adhering to these guidelines will help establish consistency at the page level and when constrained within another element or component.
+
+## Actions
+
+### Button sets
```
Do you anticipate more content being added to the "Actions" section? If not, it may help reduce the headings displayed in the sidecar to decide between "Actions" or "Button sets" as the singular heading for this content.

heatherlarsen reviewed (commented):
This was much easier to scan through; thank you for putting in the time and effort to re-organize the content! 

I added a few more minor suggestions, but I'd like us to get alignment on the "Required and optional" section before we move this forward. I see you added it to our design agenda for tomorrow, so thank you for doing that. Getting super close to wrapping this one up! ✨

jorytindall commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,269 @@
+
+!!! Dont
+
+Don’t use dividers at the end of the last section between the fields and the button set or actions. Instead, use 32px of space.
```
I revised this a bit, I think the images were actually what were lacking clarity so I updated them and broke out the "instead" statement into it's own block.

heatherlarsen commented on website/docs/patterns/form-patterns/partials/guidelines/guidelines.md:
```diff
@@ -0,0 +1,249 @@
+
+For complex forms, indicate **required** fields. This is the most explicit and transparent method and ensures users don't have to make assumptions.
+
+For shorter, simpler forms, indicate optional fields instead.
```
```suggestion
For shorter, simpler forms, indicate **optional** fields instead.
```

heatherlarsen reviewed (approved):
Just one minor consistency thing. This is good to go! :shipit: ✨
